docs/how-to/code-signing/android.rst

```diff
@@ -0,0 +1,77 @@
+the authoritative key for your app. This approach is called `app signing by Google Play
+<https://support.google.com/googleplay/android-developer/answer/7384423>`__.
+
+You will need to generate a key on your development workstation to sign an app package
+before sending it to the Google Play store. If you use app signing by Google Play,
+the key on your workstation is called the upload key.
+
+Generate a key
+--------------
+
+You will need to decide where to store the upload key. A good default is to use one
+keystore file per app you are creating and to store it in the ``.android``
+folder within your home folder. The folder is automatically created by the Android
+tools, but if it doesn't exist, we create it.
+
+We recommend using a separate keystore file per app. Below, we use the
+**upload-key-helloworld.jks** filename. This assumes you are building an app
+called "Hello World"; we informally convert it to a lowercase string
+without spaces, **helloworld**.
+
+Try not to lose this key; make backups if needed. If you do lose this key,
+you can contact Google Play support to reset it. See `Lost or compromised
+upload key
+<https://support.google.com/googleplay/android-developer/answer/7384423#reset>`__.
+If you choose not to use app signing by Google Play, it is absolutely essential
+that you not lose this key. For this reason, we recommend using app signing by
+Google Play.
+
+.. tabs::
+
+  .. group-tab:: macOS
+
+    .. code-block:: bash
+
+      $ mkdir -p ~/.android
+      $ ~/.briefcase/tools/java/Contents/Home/bin/keytool -keyalg RSA -deststoretype pkcs12 -genkey -v -storepass android -keystore ~/.android/upload-key-helloworld.jks -keysize 2048 -dname "cn=Upload Key" -alias upload-key -validity 10000
+
+  .. group-tab:: Linux
+
+    .. code-block:: bash
+
+      $ mkdir -p ~/.android
+      $ ~/.briefcase/tools/java/bin/keytool -keyalg RSA -deststoretype pkcs12 -genkey -v -storepass android -keystore ~/.android/upload-key-helloworld.jks -keysize 2048 -dname "cn=Upload Key" -alias upload-key -validity 10000
+
+  .. group-tab:: Windows
+
+    .. code-block:: doscon
+
+      C:\...> IF not exist %HOMEPATH%\.android mkdir %HOMEPATH%\.android
+      C:\...> %HOMEPATH%\.briefcase\tools\java\bin\keytool.exe -keyalg RSA -deststoretype pkcs12 -genkey -v -storepass android -keystore %HOMEPATH%\.android\upload-key-helloworld.jks -keysize 2048 -dname "cn=Upload Key" -alias upload-key -validity 10000
+
+
+This creates a 2048-bit key and stores it in a Java keystore located in the
+``.android`` folder within your home folder. Since the key's purpose is to be
+used as an upload key for the Google Play store, we call the key "upload-key".
+
+We use a password of **android**. This is the `default password for common
+Android keystores <https://developers.google.com/android/guides/client-auth>`__.
+You can change the password if you like. It is more important to limit who
+has access to the keystore file than to change the password.
```

Comment:
Any reference on “use the default password" as best practice? The general philosophy makes sense, but if we're going to recommend using a default password, I want to be sure we're giving good advice.

Reply:
Unfortunately, no -- this page here talks about creating and confirming a secure password for your keystore -- https://developer.android.com/studio/publish/app-signing -- followed by a different password for the key.

This is IMHO near-total insanity.

A Java keystore does nearly nothing, and the Java keystore password is approximately useless. See https://gist.github.com/zach-klippenstein/4631307 for demo code that shows how to change the password on a keystore without knowing the old password.

In our case, I'm using the PKCS #12 format for keystores, which may or may not have that specific problem, but it adds no value to the key password if the keystore only contains one key, which is how we suggest doing it.

As for the key's password... this is a classic ambiguity in current crypto recommendations. I'll spare you my standard rant for now, but I'm also happy to offer it to you if useful. My suggestion would be that the password adds no value if you're not sharing the key, and if you are sharing the key, do you really think the password you picked is solid enough to protect the key? If so, then you surely need to put it somewhere safe, because you surely can't remember it... in which case you have now opened a massive can of worms. Oops, I gave you 50% of the standard rant anyway.